pallets/merkle/src/lib.rs

```diff
@@ -355,19 +357,35 @@ impl<T: Config> Group<T::AccountId, T::BlockNumber, T::GroupId> for Module<T> {
 		leaf_com: Commitment,
 		path: Vec<(Commitment, Commitment)>,
 		r_com: Commitment,
-		nullifier: Data,
+		nullifier_com: Commitment,
+		nullifier_hash: Data,
 		proof_bytes: Vec<u8>
 	) -> Result<(), dispatch::DispatchError> {
 		let h = default_hasher();
 		let mut verifier_transcript = Transcript::new(b"zk_membership_proof");
 		let mut verifier = Verifier::new(&mut verifier_transcript);
 
 		let var_leaf = verifier.commit(leaf_com.0);
-		let var_s = verifier.commit(r_com.0);
-		let leaf_lc =
-			Data::constrain_verifier(&mut verifier, &pc_gens, var_s.into(), nullifier.0.into(), &h);
+		let var_r = verifier.commit(r_com.0);
+		let var_nullifier = verifier.commit(nullifier_com.0);
+		let leaf_lc = Data::constrain_verifier(
+			&mut verifier,
+			&pc_gens,
+			var_r.into(),
+			var_nullifier.into(),
+			&h
+		);
 		// Commited leaf value should be the same as calculated
 		verifier.constrain(leaf_lc - var_leaf);
+		// committed nullifier into a hash should match hash of nullifier
+		let nullifier_hash_lc = Data::constrain_verifier(
+			&mut verifier,
+			&pc_gens,
+			var_nullifier.into(),
+			var_nullifier.into(),
+			&h
+		);
+		verifier.constrain(nullifier_hash_lc - LinearCombination::from(nullifier_hash.0));
```

Comment:
We can subtract Scalar from LinearCombination, no need for conversion

Reply:
Although bulletproofs is missing some of the arithmetic impls, e.g. for references of LinearCombinations